### PR TITLE
feat: Improving filter parsing errors

### PR DESCRIPTION
**Pull request by yhakbar**
## Description

Improves the error messages returned when users use invalid filter queries.

This is what parse errors look like right now:
current-parse-errors

This is what they've been updated to look like:
updated-parse-errors

Generally, hints are only suggested when there's a reliable suggestion that can fix the user's filter query, and the error message tries to reference the exact location that's broken in the filter query when there's an error.

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [ ] I authored this code entirely myself
- [x] I am submitting code based on open source software (e.g. MIT, MPL-2.0, Apache)]
- [x] I am adding or upgrading a dependency or adapted code and confirm it has a compatible open source license
- [x] Update the docs.
- [x] Run the relevant tests successfully, including pre-commit checks.
- [x] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes (draft)

Added / Removed / Updated [X].

## Release Notes

* **New Features**
  * Enhanced filter parsing error diagnostics with contextual hints, precise position markers, and formatted error messages for improved troubleshooting.
  * Added color-aware error output formatting for filter parsing errors when supported.

* **Tests**
  * Added comprehensive test suite for filter error diagnostics and hint generation.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: internal/filter/diagnostic.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+//	    [main...]
+//	            ^ Expected second Git reference after '...'
+//
+//	 hint: Git filters with '...' require a reference on each side. e.g. '[main...HEAD]'
+//
+// ```
+func FormatDiagnostic(err *ParseError, filterIndex int, useColor bool) string {
+	var sb strings.Builder
+
+	fmt.Fprintf(&sb, "Filter parsing error: %s\n", err.Title)
```
Comment (denis256) [inline internal/filter/diagnostic.go:35]: Error not handled

File: internal/filter/hints.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	// These are errors that don't have obvious hints that can be offered.
+	case ErrorCodeUnknown:
+		return ""
+	}
+
+	return ""
+}
+
+// GetHints returns context-aware hints for a parse error (legacy, returns slice for compatibility).
+func GetHints(code ErrorCode, token, query string, position int) []string {
```
Comment (denis256) [inline internal/filter/hints.go]: Looks like GetHints() is not used

File: internal/filter/diagnostic.go (hunk 1 above)
Comment (yhakbar) [reply internal/filter/diagnostic.go:35]: There's no point. `sb.Write` always returns nil:
https://pkg.go.dev/strings#Builder.Write

If we were a project that liked asserts, I could assert that this is true, but I don't really see the value here, and we probably want fuzzing or some more advance testing before we start adding in that kind of aggressive negative space logic.

File: internal/filter/hints.go (hunk 2 above)
Comment (yhakbar) [reply internal/filter/hints.go]: Removed.
